### PR TITLE
drivers/at86rf215: implement MR-FSK

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This adds the MR-FSK modulation to the `at86rf215` driver.
You can configure all possible combinations of settings, grouping those settings to "FSK Modes" as per 802.15.4 is out of scope of the driver and should become part of the new 802.15.4 sub-MAC so it can be shared with other drivers implementing MR-FSK.

The same goes for channel spacing. @tinstructor added the option to toggle between 200 kHz and 400 kHz channel spacing (600 kHz is omitted for now), but again this is hardly useful isolation and should not be specific to MR-FSK.

Again, sorting this out properly would have to be done in common code (and deriving the number of channels from frequency band with / channel spacing if possible).

@jeandudey do those settings also make sense for `cc26x2_cc13x2_rf`?

### Testing procedure

You should now be able to configure the `mr-fsk` phy mode on boards with the `at86rf215` driver.

```
2020-09-07 00:37:02,866 #  ifconfig 6 set phy mr-fsk
2020-09-07 00:37:05,846 # Iface  6  HWaddr: 04:8F  Channel: 0  NID: 0x23  PHY: MR-FSK 
2020-09-07 00:37:05,854 #            modulation index: 1  2-FSK  symbol rate: 50 kHz  FEC: none  BW: 400kHz 
2020-09-07 00:37:05,858 #           Long HWaddr: 1A:F9:8F:5F:30:5D:04:8F 
2020-09-07 00:37:05,865 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2020-09-07 00:37:05,871 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:2022  MTU:1280  HL:64  RTR  
2020-09-07 00:37:05,873 #           6LO  IPHC  
2020-09-07 00:37:05,876 #           Source address length: 8
2020-09-07 00:37:05,879 #           Link type: wireless
2020-09-07 00:37:05,884 #           inet6 addr: fe80::18f9:8f5f:305d:48f  scope: link  VAL
2020-09-07 00:37:05,887 #           inet6 group: ff02::2
2020-09-07 00:37:05,890 #           inet6 group: ff02::1
2020-09-07 00:37:05,893 #           inet6 group: ff02::1:ff5d:48f
```

You should be able to set modulation order, symbol rate, forward error correction and modulation index.

```
2020-09-07 00:39:25,712 #  ifconfig 6 set srate 200
2020-09-07 00:39:39,103 #  ifconfig 6 set mord 4
2020-09-07 00:39:56,624 #  ifconfig 6 set fec rsc
2020-09-07 00:40:21,936 #  ifconfig 6 set modulation_index 3/4

2020-09-07 00:40:25,611 # Iface  6  HWaddr: 04:8F  Channel: 0  NID: 0x23  PHY: MR-FSK 
2020-09-07 00:40:25,619 #            modulation index: 3/4  4-FSK  symbol rate: 200 kHz  FEC: RSC  BW: 400kHz 
2020-09-07 00:40:25,623 #           Long HWaddr: 1A:F9:8F:5F:30:5D:04:8F 
2020-09-07 00:40:25,630 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2020-09-07 00:40:25,636 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:2022  MTU:1280  HL:64  RTR  
2020-09-07 00:40:25,638 #           6LO  IPHC  
2020-09-07 00:40:25,641 #           Source address length: 8
2020-09-07 00:40:25,644 #           Link type: wireless
2020-09-07 00:40:25,649 #           inet6 addr: fe80::18f9:8f5f:305d:48f  scope: link  VAL
2020-09-07 00:40:25,652 #           inet6 group: ff02::2
2020-09-07 00:40:25,655 #           inet6 group: ff02::1
2020-09-07 00:40:25,658 #           inet6 group: ff02::1:ff5d:48f
```

Communication should still work.
```
2020-09-07 00:43:25,820 #  ping6 ff02::1%6
2020-09-07 00:43:25,837 # 12 bytes from fe80::18f9:8f5f:305d:48f%6: icmp_seq=0 ttl=64 rssi=-30 dBm time=11.322 ms
2020-09-07 00:43:26,829 # 12 bytes from fe80::18f9:8f5f:305d:48f%6: icmp_seq=1 ttl=64 rssi=-30 dBm time=11.319 ms
2020-09-07 00:43:27,853 # 12 bytes from fe80::18f9:8f5f:305d:48f%6: icmp_seq=2 ttl=64 rssi=-30 dBm time=28.797 ms
2020-09-07 00:43:27,853 # 
2020-09-07 00:43:27,854 # --- ff02::1 PING statistics ---
2020-09-07 00:43:27,856 # 3 packets transmitted, 3 packets received, 0% packet loss
2020-09-07 00:43:27,868 # round-trip min/avg/max = 11.319/17.146/28.797 ms
```
However, not all combinations will produce valid results.

### Issues/PRs references

split off #12128
#13295 implements a driver for a different MR-FSK radio

### CODE REVIEW

**Comment by fjmolinas** [comment]
> @jeandudey do those settings also make sense for cc26x2_cc13x2_rf?

@jeandudey if you can give the common `mr-fsk` code a look it would be much appreciated!

**Comment by benpicco** [comment]
Rebased to re-add RPC

**Comment by jeandudey** [comment]
Hi! Sorry for the late response. Yesterday I took a look on the code and it seems very well written and everything here applies on the other driver as much on the netdev driver configuration.

Another options I saw on the AT86RF215 datasheet that are configurable too on the `cc26x2_cc13x2_rf` driver was the Data Whitening option (_phyFSKScramblePSDU_) which is enabled by default for 802.15.4g and it's fixed by TI :-( and could affect interoperability between the two transceivers, the Preamble Length (_phyFSKPreambleLength_) which is not mandatory to have as a configuration option but could be good to have the same value on both radios as default, the FCS type e.g. for 16-bit or 32-bit CRC which in theory both modes are supported on `cc26x2_cc13x2_rf` although I have only used the 32-bit CRC.

Overall the only options that may affect interop are Data Whitening and the FCS, the SFD is another factor that could affect but I may assume that AT86RF215 uses the default values the IEEE 802.15.4g standard provides (see https://github.com/RIOT-OS/RIOT/pull/13295/commits/85423bd28b37f07f0e3347aaac6da88082ccb9eb) when using FEC or not, although it could be nice if could be selected between _phyMRFSKSFD=0_ and _phyMRFSKSFD=1_.

**Comment by benpicco** [comment]
> Another options I saw on the AT86RF215 datasheet that are configurable too on the cc26x2_cc13x2_rf driver was the Data Whitening option (phyFSKScramblePSDU) which is enabled by default for 802.15.4g and it's fixed by TI 

That can be enabled / disabled by toggling the `DW` bit the `FSKPHRTX` register.
It's enabled by default and I didn't touch it, but I can make it configurable via `ifconfig` if that's useful.

> the Preamble Length (phyFSKPreambleLength) which is not mandatory to have as a configuration option but could be good to have the same value on both radios as default, 

I got the minimum preamble length for a symbol rate from Table 6-64 of the at86rf215 data sheet.
That is the minimal preamble length on RX to recognize the start of a frame.
On TX it will send 8x that length. It has to send more preamble than the minimal length to allow for robust recognition, but I don't remember how I arrived at the 8x figure.

> the FCS type e.g. for 16-bit or 32-bit CRC which in theory both modes are supported on `cc26x2_cc13x2_rf` although I have only used the 32-bit CRC.

Yea I only ever used 16-bit FCS like the old modes did.
IIUC it should be possible to receive frames of both types and the transceiver will tell them apart by a bit in the header.

Again, I can add a config option for that. That would apply to all 802.15.4g modes.

> it could be nice if could be selected between phyMRFSKSFD=0 and phyMRFSKSFD=1.

Do you also want to set the content of the start of frame delimiter? :D 
But I see that they have different defaults, probably as per standard.
What is the meaning of the two SFDs? When would you chose one over the other?

I already

**Comment by jeandudey** [comment]
>That can be enabled / disabled by toggling the DW bit the FSKPHRTX register.
>It's enabled by default and I didn't touch it, but I can make it configurable via ifconfig if that's useful.

If enabled by default it could then be leaved as is, it could only make sense to disable DW for testing purposes.

>Do you also want to set the content of the start of frame delimiter? :D
But I see that they have different defaults, probably as per standard.

Right now it isn't needed until some rare  802.15.4g transceiver appears with rare configurations :P so I wouldn't bother with it.

>What is the meaning of the two SFDs? When would you chose one over the other?

The SFD changes between 2-FSK and 4-FSK (2 octets vs 4 octets), and when using FEC or not. Selecting between one group or another it's just a matter of taste, and only mandatory for Common Signaling Mode to use a _phyMRFSKSFD=0_, but so far I haven't seen CSM out there.

> IIUC it should be possible to receive frames of both types and the transceiver will tell them apart by a bit in the header.

Mine bad, revisited the configurations for `cc26x2_cc13x2_rf` and it should receive both types of packets depending on the bit in the PHR and uses a 16-bit CRC for FCS so it should be fine as is.

**Comment by benpicco** [comment]
Ok, I'll keep the amount of ifconfig tunables at a minimum then :)
They take up considerable amount of space and we don't have a good way to make them optional yet.
And ideally the user wouldn't have to deal with those settings anyway.

> > What is the meaning of the two SFDs? When would you chose one over the other?
> 
> The SFD changes between 2-FSK and 4-FSK (2 octets vs 4 octets), and when using FEC or not.

Hm both config registers for SFD0 and SFD1 are 2 octets - do you suggest both are used for 4-FSK and FEC?

> revisited the configurations for `cc26x2_cc13x2_rf` and it should receive both types of packets depending on the bit in the PHR and uses a 16-bit CRC for FCS

Ah I reckon `cc26x2_cc13x2_rf` doesn't do the CRC verification in hardware then?

**Comment by jeandudey** [comment]
>Ah I reckon cc26x2_cc13x2_rf doesn't do the CRC verification in hardware then?

It was me :sweat_smile: that didn't verified how it handled the FCS bit on the PHR, it will work with both and has to be specified on TX manually by setting/clearing the FCS bit on the PHR which now defaults to 16-bit CRC. It indeed handles CRC verification on hardware.

File: drivers/at86rf215/at86rf215_fsk.c [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+        return 8;
+    case FSK_SRATE_400K:
+        return 10;
+    }
+
+    return 0;
+}
+
+/* fsk modulation indices / 8 */
+static const uint8_t fsk_mod_idx[] = {
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_fsk.c:312]:
```suggestion
static const uint8_t _fsk_mod_idx[] = {
```

File: drivers/at86rf215/at86rf215_fsk.c [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+    3, 4, 6, 8, 10, 12, 14, 16
+};
+
+/* FSK modulation scale / 8 */
+static const uint8_t fsk_mod_idx_scale[] = {
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_fsk.c:317]:
```suggestion
static const uint8_t _fsk_mod_idx_scale[] = {
```

File: drivers/at86rf215/at86rf215_fsk.c [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+/* also used by at86rf215_netdev.c */
+const uint8_t at86rf215_fsk_channel_spacing_25kHz[] = {
+    [FSK_CHANNEL_SPACING_200K] = 8,
+    [FSK_CHANNEL_SPACING_400K] = 16,
+};
+
+/* IEEE Std 802.15.4™-2015
+ * Table 10-10—Channel numbering for SUN PHYs,
+ * index is channel spacing */
+static const uint16_t chan_center_freq0_subghz_25khz[] = {
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_fsk.c:48]:
```suggestion
static const uint16_t _chan_center_freq0_subghz_25khz[] = {
```

File: drivers/at86rf215/at86rf215_fsk.c [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+    [FSK_CHANNEL_SPACING_200K] = 863125U / 25,
+    [FSK_CHANNEL_SPACING_400K] = 863225U / 25
+};
+
+/* IEEE Std 802.15.4™-2015
+ * Table 10-10—Channel numbering for SUN PHYs,
+ * index is channel spacing */
+static const uint16_t chan_center_freq0_24ghz_25khz[] = {
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_fsk.c:56]:
```suggestion
static const uint16_t _chan_center_freq0_24ghz_25khz[] = {
```

File: drivers/at86rf215/at86rf215_fsk.c [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+    [FSK_CHANNEL_SPACING_200K] = (2400200U - CCF0_24G_OFFSET) / 25,
+    [FSK_CHANNEL_SPACING_400K] = (2400400U - CCF0_24G_OFFSET) / 25
+};
+
+/* Table 6-57, index is symbol rate */
+static const uint8_t FSKPE_Val[3][6] = {
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_fsk.c:62]:
```suggestion
static const uint8_t _FSKPE_Val[3][6] = {
```

File: drivers/at86rf215/at86rf215_netdev.c [hunk 6]
```diff
@@ -393,7 +407,37 @@ static int _get(netdev_t *netdev, netopt_t opt, void *val, size_t max_len)
...
 
+        case NETOPT_MR_FSK_MODULATION_ORDER:
+            assert(max_len >= sizeof(int8_t));
+            *((int8_t *)val) = 2 + 2*at86rf215_FSK_get_mod_order(dev);
+            res = max_len;
+            break;
+
+        case NETOPT_MR_FSK_SRATE:
+            assert(max_len >= sizeof(uint16_t));
+            *((uint16_t *)val) = 10 * at86rf215_fsk_srate_10kHz[at86rf215_FSK_get_srate(dev)];
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_netdev.c:425]: Magic number?

File: drivers/at86rf215/at86rf215_netdev.c [hunk 7]
```diff
@@ -393,7 +407,37 @@ static int _get(netdev_t *netdev, netopt_t opt, void *val, size_t max_len)
...
+#ifdef MODULE_NETDEV_IEEE802154_MR_FSK
+        case NETOPT_MR_FSK_MODULATION_INDEX:
+            assert(max_len >= sizeof(int8_t));
+            *((int8_t *)val) = at86rf215_FSK_get_mod_idx(dev);
+            res = max_len;
+            break;
 
+        case NETOPT_MR_FSK_MODULATION_ORDER:
+            assert(max_len >= sizeof(int8_t));
+            *((int8_t *)val) = 2 + 2*at86rf215_FSK_get_mod_order(dev);
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_netdev.c:419]: magic number?

File: drivers/at86rf215/at86rf215_netdev.c [hunk 8]
```diff
@@ -601,11 +645,88 @@ static int _set(netdev_t *netdev, netopt_t opt, const void *val, size_t len)
...
+
+        case NETOPT_MR_FSK_MODULATION_ORDER:
+            if (at86rf215_get_phy_mode(dev) != IEEE802154_PHY_MR_FSK) {
+                return -ENOTSUP;
+            }
+
+            if (*(uint8_t *)val != 2 && *(uint8_t *)val != 4) {
+                res = -ERANGE;
+            } else {
+                at86rf215_FSK_set_mod_order(dev, *(uint8_t *)val >> 2);
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_netdev.c:684]: Magic numbers?

File: drivers/at86rf215/at86rf215_netdev.c [hunk 9]
```diff
@@ -601,11 +645,88 @@ static int _set(netdev_t *netdev, netopt_t opt, const void *val, size_t len)
...
+            break;
+
+        case NETOPT_MR_FSK_SRATE:
+            if (at86rf215_get_phy_mode(dev) != IEEE802154_PHY_MR_FSK) {
+                return -ENOTSUP;
+            }
+
+            res = _get_best_match(at86rf215_fsk_srate_10kHz,
+                                  FSK_SRATE_400K + 1, *(uint16_t *)val / 10);
+            if (at86rf215_FSK_set_srate(dev, res) == 0) {
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_netdev.c:696]: Magic numbers?

File: drivers/at86rf215/at86rf215_netdev.c [hunk 10]
```diff
@@ -601,11 +645,88 @@ static int _set(netdev_t *netdev, netopt_t opt, const void *val, size_t len)
...
+
+        case NETOPT_CHANNEL_SPACING:
+            if (at86rf215_get_phy_mode(dev) != IEEE802154_PHY_MR_FSK) {
+                return -ENOTSUP;
+            }
+
+            res = _get_best_match(at86rf215_fsk_channel_spacing_25kHz,
+                                  FSK_CHANNEL_SPACING_400K + 1, *(uint16_t *)val / 25);
+            if (at86rf215_FSK_set_channel_spacing(dev, res) == 0) {
+                res = 25 * at86rf215_fsk_channel_spacing_25kHz[res];
```
Comment (fjmolinas) [inline drivers/at86rf215/at86rf215_netdev.c:724]: Magic numbers?

File: drivers/at86rf215/include/at86rf215_internal.h [hunk 11]
```diff
@@ -46,8 +46,8 @@ extern "C" {
 /**
  * This is used to calculate the csma backoff based on the bitrate.
  */
-/** 20 symbols is the std period length */
-#define AT86RF215_BACKOFF_PERIOD_IN_SYMBOLS (20U)
+/** 8 symbols is the std period length */
+#define AT86RF215_BACKOFF_PERIOD_IN_SYMBOLS (8U)
```
Comment (fjmolinas) [inline drivers/at86rf215/include/at86rf215_internal.h:50]: Why this change?

File: sys/shell/commands/sc_gnrc_netif.c [hunk 12]
```diff
@@ -96,6 +96,42 @@ static void str_toupper(char *str)
     }
 }
 
+__attribute__ ((unused))
+static uint8_t gcd(uint8_t a, uint8_t b)
```
Comment (fjmolinas) [inline sys/shell/commands/sc_gnrc_netif.c:100]: there is one in `frac.c` can't you use that?

**Review by fjmolinas — COMMENTED** [review]
Some nitpicks, otherwise looks good.

**Comment by fjmolinas** [comment]
I played around changing configurations and pinging everything seems OK.

**Comment by fjmolinas** [comment]
@jeandudey are you fine with the general FSK stuff?

File: drivers/at86rf215/include/at86rf215_internal.h (hunk 11 above)
Comment (benpicco) [reply drivers/at86rf215/include/at86rf215_internal.h:50]: Meh, both are wrong :weary: 

image

with 

#### aTurnaroundTime

image

#### aCCATime

> For the SUN PHYs other than SUN O-QPSK, the duration of 8 symbol periods, as defined in 6.1. For the SUN O-QPSK PHY, this value is defined in Table 21-24. For all other PHYs, the duration of 8 symbol periods.

image

File: drivers/at86rf215/at86rf215_netdev.c (hunk 8 above)
Comment (benpicco) [reply drivers/at86rf215/at86rf215_netdev.c:684]: 4-FSK -> 1
2-FSK -> 0

I added a comment

File: sys/shell/commands/sc_gnrc_netif.c (hunk 12 above)
Comment (benpicco) [reply sys/shell/commands/sc_gnrc_netif.c:100]: yea that just looked so complex and I didn't need speed here 😅

**Comment by jeandudey** [comment]
> @jeandudey are you fine with the general FSK stuff?

Yes, everything seems very well to me in regards with the `netopt` options and will map 1to1 to the `cc26x2_cc13x2_rf` driver.

File: drivers/at86rf215/at86rf215_netdev.c (hunk 9 above)
Comment (benpicco) [reply drivers/at86rf215/at86rf215_netdev.c]: added a comment.
(Not sure if it makes sense to use 2 and 4 in NETOPT since there is no '3-FSK' or '17-FSK', but it sure makes `ifconfig` easier that way)

File: drivers/at86rf215/include/at86rf215_internal.h (hunk 11 above)
Comment (benpicco) [reply drivers/at86rf215/include/at86rf215_internal.h]: should be fixed now.

File: sys/shell/commands/sc_gnrc_netif.c (hunk 12 above)
Comment (fjmolinas) [reply sys/shell/commands/sc_gnrc_netif.c:100]: You lost me a bit a here, where would this database/table be used?

Comment (benpicco) [reply sys/shell/commands/sc_gnrc_netif.c:100]: As with WiFi different world regions have varying regulations as to what portions of the spectrum you may use and how you may use it. And while the 2.4 GHz ISM band can be used everywhere, for sub-GHz different countries use different frequencies for ISM and the [standard](https://sci-hub.se/https://ieeexplore.ieee.org/document/9144691) prescribes different modulation parameters for those frequencies.
 
image

image

So in the end you would really want to be able to "set region EU, set fsk_mode 3" and not worry about the details.

But that should be handled outside the driver so other ~~MR~~SUN-FSK like `cc26x2_cc13x2_rf` can also benefit from it.

**Comment by fjmolinas** [comment]
Please squash @benpicco !

**Review by fjmolinas — APPROVED** [review]
ACK!
